x/stakeibc/keeper/callbacks.go

```diff
@@ -176,3 +182,166 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
 
 	return nil
 }
+
+// get a validator and its index from a list of validators, by address
+func getValidator(validators []*types.Validator, address string) (types.Validator, int64, bool) {
+	for i, v := range validators {
+		if v.Address == address {
+			return *v, int64(i), true
+		}
+	}
+	return types.Validator{}, 0, false
+}
+
+// ValidatorCallback is a callback handler for validator queries.
+func ValidatorExchangeRateCallback(k Keeper, ctx sdk.Context, args []byte, query icqtypes.Query) error {
+	zone, found := k.GetHostZone(ctx, query.GetChainId())
+	if !found {
+		return fmt.Errorf("no registered zone for chain id: %s", query.GetChainId())
+	}
+	queriedValidator := stakingtypes.Validator{}
+	err := k.cdc.Unmarshal(args, &queriedValidator)
+	if err != nil {
+		k.Logger(ctx).Error(fmt.Sprintf("unable to unmarshal queriedValidator info for zone %s, err: %s", zone.ChainId, err.Error()))
+		return err
+	}
+	k.Logger(ctx).Info(fmt.Sprintf("ValidatorCallback: zone %v queriedValidator %v", zone.ChainId, queriedValidator))
+
+	// ensure ICQ can be issued now! else fail the callback
+	valid, err := k.IsWithinBufferWindow(ctx)
+	if err != nil {
+		return err
+	} else if !valid {
+		return nil
+	}
+
+	// set the validator's conversion rate
+	v, i, found := getValidator(zone.Validators, queriedValidator.OperatorAddress)
+	if !found {
+		return fmt.Errorf("no registered validator for address: %s", queriedValidator.OperatorAddress)
+	}
+	// get the stride epoch number
+	strideEpochTracker, found := k.GetEpochTracker(ctx, epochtypes.STRIDE_EPOCH)
+	if !found {
+		k.Logger(ctx).Error("failed to find stride epoch")
+		return sdkerrors.Wrapf(sdkerrors.ErrNotFound, "no epoch number for epoch (%s)", epochtypes.STRIDE_EPOCH)
+	}
+	// converting 1.0 gives us the exchange rate to later use in the next CB
+	v.InternalExchangeRate = &types.ValidatorExchangeRate{
+		InternalTokensToSharesRate: queriedValidator.TokensFromShares(sdk.NewDec(1.0)),
+		EpochNumber:                strideEpochTracker.GetEpochNumber(),
```

Comment:
given we don't have TTLs, this is a bit misleading in that `EpochNumber` is "the epoch the query returned in", not the epoch it was issued in. just noting that the exchange rate might have been calculated at a previous `EpochNumber`

Reply:
> few minor comments, after resolving those good to merge I think. it does seem like there are a lot of commits towards the end of this PR, what are those? do you want a review on those as well?
> 
> we should prioritize adding TTLs, the edge cases are hard to reason about without them

TTLs are up next!